contracts/LockedNORI.sol

```diff
@@ -40,127 +63,339 @@ contract LockedNORI is ERC777Upgradeable, ERC20PresetMinterPauserUpgradeable, ER
+    //   console.log("grant.lockupSchedule.availableAmount(block.timestamp)",grant.lockupSchedule.availableAmount(block.timestamp));
+    //   console.log("grant.claimedAmount",grant.claimedAmount);
+      // If any tokens have been revoked then the schedule (which doesn't get updated) could
+      // return more than the total grant amount.
       return
-        grant.lockupSchedule.availableAmount(block.timestamp) -
-        grant.claimedAmount;
+        MathUpgradeable.min(
+          MathUpgradeable.min(
+            grant.vestingSchedule.availableAmount(block.timestamp),
+            grant.lockupSchedule.availableAmount(block.timestamp)
+          ),
+          grant.grantAmount
+        ) - grant.claimedAmount;
     }
-    return depositedBalanceOf(account);
+    return 0;
   }
 
-  function balanceOf(address account) public view override(ERC20Upgradeable, ERC777Upgradeable) returns (uint256) {
-    return unlockedBalanceOf(account);
+  /**
+   * @dev balanceOf: Total amount granted to the user less any previously claimed amount
+   */
+  function balanceOf(address account)
+    public
+    view
+    override(ERC20Upgradeable, ERC777Upgradeable)
+    returns (uint256)
+  {
+    return ERC777Upgradeable.balanceOf(account);
   }
 
-  function _beforeTokenTransfer(
-    address from,
-    address to,
-    uint256 amount
-  ) internal override(ERC20PresetMinterPauserUpgradeable, ERC20Upgradeable) {
-    if (to == address(0)) {
-      uint256 availableBalance = balanceOf(from);
-      uint256 wrappedBalance = super.balanceOf(from);
-      console.log(
-        "amount: %d  available: %d  wrapped: %d",
-        amount,
-        availableBalance,
-        wrappedBalance
-      );
-      require(amount <= availableBalance, "Withdrawl amount unavailable");
+  /**
+   * @dev getGrant: Returns all governing settings on a grant.
+   */
+  function getGrant(address account)
+    public
+    view
+    returns (
+      uint256 amount,
+      uint256 startTime,
+      uint256 vestEndTime,
+      uint256 unlockEndTime,
+      uint256 cliff1Time,
+      uint256 cliff2Time,
+      uint256 vestCliff1Amount,
+      uint256 vestCliff2Amount,
+      uint256 unlockCliff1Amount,
+      uint256 unlockCliff2Amount,
+      uint256 claimedAmount,
+      uint256 originalAmount
+    )
+  {
+    TokenGrant storage grant = grants[account];
+    return (
+      grant.grantAmount,
+      grant.lockupSchedule.startTime,
+      grant.vestingSchedule.endTime,
+      grant.lockupSchedule.endTime,
+      grant.lockupSchedule.cliffs[0].time,
+      grant.lockupSchedule.cliffs[1].time,
+      grant.vestingSchedule.cliffs[0].amount,
+      grant.vestingSchedule.cliffs[1].amount,
+      grant.lockupSchedule.cliffs[0].amount,
+      grant.lockupSchedule.cliffs[1].amount,
+      grant.claimedAmount,
+      grant.originalAmount
+    );
+  }
+
+  function _isTokenGranter(address account) internal view returns (bool) {
```

Comment:
what is this and why is it needed in addition to `hasRole(TOKEN_GRANTER_ROLE)`

Reply:
removed